sonic_platform_base/fan_base.py

```diff
@@ -0,0 +1,100 @@
+#
+# fan_base.py
+#
+# Abstract base class for implementing a platform-specific class with which
+# to interact with a fan module in SONiC
+#
+
+try:
+    import abc
+    from . import device_base
+except ImportError as e:
+    raise ImportError(str(e) + " - required module not found")
+
+
+class FanBase(device_base.DeviceBase):
+    """
+    Abstract base class for interfacing with a fan module
+    """
+    __metaclass__ = abc.ABCMeta
+    
+    # Possible fan directions
+    FAN_DIRECTION_INTAKE = "intake"
+    FAN_DIRECTION_EXHAUST = "exhaust"
+
+    # Possible fan status LED colors
+    STATUS_LED_COLOR_GREEN = "green"
+    STATUS_LED_COLOR_RED = "red"
+    STATUS_LED_COLOR_OFF = "off"
+
+    @abc.abstractmethod
+    def get_direction(self):
+        """
+        Retrieves the direction of fan
+
+        Returns:
+            A string, either FAN_DIRECTION_INTAKE or FAN_DIRECTION_EXHAUST
+            depending on fan direction
+        """
+        return None
+
+    @abc.abstractmethod
+    def get_speed(self):
+        """
+        Retrieves the speed of fan as a percentage of full speed
+
+        Returns:
+            An integer, the percentage of full fan speed, in the range 0 (off)
+                 to 100 (full speed)
+        """
+        return 0
+
+    @abc.abstractmethod
+    def get_target_speed(self):
+        """
+        Retrieves the target (expected) speed of the fan
+
+        Returns:
+            An integer, the percentage of full fan speed, in the range 0 (off)
+                 to 100 (full speed)
+        """
+        return 0
+
+    @abc.abstractmethod
```

Comment:
Does this method have to be abstract or a default implementation that is always overridable would do?

Reply:
Assuming tolerance is most commonly implemented as a static attribute, we could define a `tolerance` member and have that returned here as the default implementation. Would that suffice for you?